File: week12/23/prob4_implement_queue_using_stacks/prob4.py
class MyQueue:
    def __init__(self):
        self.stack1 = []  # Used for push operations
        self.stack2 = []  # Used for pop/peek operations

    def push(self, x: int) -> None:
        self.stack1.append(x)

    def pop(self) -> int:
        self._move()
        return self.stack2.pop()

    def peek(self) -> int:
        self._move()
        return self.stack2[-1]

    def empty(self) -> bool:
        return not self.stack1 and not self.stack2

    def _move(self):
        if not self.stack2:  # Only move elements if stack2 is empty
            while self.stack1:
                self.stack2.append(self.stack1.pop())

# Example usage
queue = MyQueue()
queue.push(1)
queue.push(2)
print(queue.peek())  # Output: 1
print(queue.pop())   # Output: 1
print(queue.empty()) # Output: False

File: week12/23/prob4_implement_queue_using_stacks/prob4.cpp
#include <iostream>
#include <stack>
using namespace std;

class MyQueue {
private:
    stack<int> stack1;  // Used for push operations
    stack<int> stack2;  // Used for pop/peek operations

    void move() {
        if (stack2.empty()) {
            while (!stack1.empty()) {
                stack2.push(stack1.top());
                stack1.pop();
            }
        }
    }

public:
    MyQueue() {}

    void push(int x) {
        stack1.push(x);
    }

    int pop() {
        move();
        int top = stack2.top();
        stack2.pop();
        return top;
    }

    int peek() {
        move();
        return stack2.top();
    }

    bool empty() {
        return stack1.empty() && stack2.empty();
    }
};

// Example usage
int main() {
    MyQueue queue;
    queue.push(1);
    queue.push(2);
    cout << queue.peek() << endl;  // Output: 1
    cout << queue.pop() << endl;   // Output: 1
    cout << queue.empty() << endl; // Output: 0
    return 0;
}
